src/jdk.jlink/share/classes/jdk/tools/jlink/internal/JRTArchive.java

```diff
@@ -0,0 +1,536 @@
+                int readBytes = -1;
+                while ((readBytes = is.read(buf)) != -1) {
+                    digest.update(buf, 0, readBytes);
+                }
+            }
+            byte[] actual = digest.digest();
+            return !MessageDigest.isEqual(expected, actual);
+        } catch (Exception e) {
+            throw new AssertionError("SHA-512 sum check failed!", e);
+        }
+    }
+
+    private static EntryType toEntryType(Type input) {
+        return switch(input) {
+            case CLASS_OR_RESOURCE -> EntryType.CLASS_OR_RESOURCE;
+            case CONFIG -> EntryType.CONFIG;
+            case HEADER_FILE -> EntryType.HEADER_FILE;
+            case LEGAL_NOTICE -> EntryType.LEGAL_NOTICE;
+            case MAN_PAGE -> EntryType.MAN_PAGE;
+            case NATIVE_CMD -> EntryType.NATIVE_CMD;
+            case NATIVE_LIB -> EntryType.NATIVE_LIB;
+            case TOP -> throw new IllegalArgumentException(
+                           "TOP files should be handled by ReleaseInfoPlugin!");
+            default -> throw new IllegalArgumentException(
+                           "Unknown type: " + input);
+        };
+    }
+
+    public record ResourceFileEntry(Type resType,
+                                    boolean symlink,
+                                    String hashOrTarget,
+                                    String resPath) {
+        // Type file format:
+        // '<type>|{0,1}|<sha-sum>|<file-path>'
+        //   (1)    (2)      (3)      (4)
+        //
+        // Where fields are:
+        //
+        // (1) The resource type as specified by ResourcePoolEntry.type()
+        // (2) Symlink designator. 0 => regular resource, 1 => symlinked resource
+        // (3) The SHA-512 sum of the resources' content. The link to the target
+        //     for symlinked resources.
+        // (4) The relative file path of the resource
+        private static final String TYPE_FILE_FORMAT = "%d|%d|%s|%s";
+
+        public String encodeToString() {
+            return String.format(TYPE_FILE_FORMAT,
+                                 resType.ordinal(),
+                                 symlink ? 1 : 0,
+                                 hashOrTarget,
+                                 resPath);
+        }
+
+        /**
+         *  line: <int>|<int>|<hashOrTarget>|<path>
+         *
+         *  Take the integer before '|' convert it to a Type. The second
+         *  token is an integer representing symlinks (or not). The third token is
+         *  a hash sum (sha512) of the file denoted by the fourth token (path).
+         */
+        static ResourceFileEntry decodeFromString(String line) {
+            assert !line.isEmpty();
+
+            String[] tokens = line.split("\\|", 4);
+            Type type = null;
+            int symlinkNum = -1;
+            try {
+                Integer typeInt = Integer.valueOf(tokens[0]);
+                type = Type.fromOrdinal(typeInt);
```

Comment:
Since encoding and decoding are private to this class, this class can keep a map from an ordinal to `Type` and no need to add the static `Type::fromOrdinal` method.

```
        private static final Map<Integer, Type> typeMap = Arrays.stream(Type.values())
                .collect(Collectors.toMap(Type::ordinal, Function.identity()));
```
                
```suggestion
                type = typeMap.get(typeInt);
```